### PR TITLE
Remove dependency on keycloak by using oauth of openshift and github

### PR DESCRIPTION
**Pull request by edewit**
To test create a oauth client in openshift as `system:admin`
```
oc create -f <(echo '
  kind: OAuthClient
  apiVersion: oauth.openshift.io/v1
  metadata:
    name: demo
  secret: demo
  redirectURIs:
    - "http://<your frontend hostname>/"
  grantMethod: prompt
  ')
```
Create a github oauth app: https://github.com/settings/applications/new

Fill out `.env.openshift`:
```
REACT_APP_OPENSHIFT_CLIENT_ID=demo

REACT_APP_OPENSHIFT_AUTH_URL=https://<you cluster ip>/oauth/authorize

REACT_APP_GIT_CLIENT_ID=<your app id>
REACT_APP_GIT_SECRET=<your app secret>
```

### CODE REVIEW

**Comment by fabric8cd** [comment]
This PR was deployed **with mocked data** in the following URLs for your appreciation:
- Frontend: https://1231-pr-launcher-app-launcher-frontend-fabric8-launcher.surge.sh
- Storybook: https://1231-pr-storybook-launcher-frontend-fabric8-launcher.surge.sh

File: packages/launcher-client/src/client/impl/default.launcher.client.ts [hunk 1]
```diff
@@ -156,8 +157,12 @@ export default class DefaultLauncherClient implements LauncherClient {
     if (config.gitProvider) {
       headers['X-Git-Provider'] = config.gitProvider;
     }
-    if (authorizationToken) {
+    if (typeof authorizationToken === 'string') {
       headers['Authorization'] = `Bearer ${authorizationToken}`;
+    } else if (authorizationToken) {
+      const authToken = authorizationToken as AuthorizationToken[];
+      authToken.map(t => headers[t.header] = `Bearer ${t.token}`);
+      headers['Authorization'] = 'Bearer eyJhbGciOiJIUzI1NiJ9.e30.ZRrHA1JJJW8opsbCGfG_HACGpVUMN_a9IV7pAx_Zmeo';
```
Comment (ia3andy) [inline packages/launcher-client/src/client/impl/default.launcher.client.ts]: @edewit this is until @gastaldi changes the backend right?

Comment (edewit) [reply packages/launcher-client/src/client/impl/default.launcher.client.ts]: Right, the backend always needs an `Authorization` header even if we are sending an `X-OpenShift-Authorization`

File: packages/launcher-app/src/config.ts [hunk 2]
```diff
@@ -23,11 +23,17 @@ export const authenticationMode = getEnv(process.env.REACT_APP_AUTHENTICATION, '
   || (keycloakUrl ? 'keycloak' : 'no');
 export const isKeycloakMode = authenticationMode === 'keycloak';
 
-export const keycloakConfig = isKeycloakMode ? {
+export const config = isKeycloakMode ? {
   clientId: requireEnv(process.env.REACT_APP_KEYCLOAK_CLIENT_ID, 'keycloakClientId'),
   realm: requireEnv(process.env.REACT_APP_KEYCLOAK_REALM, 'keycloakRealm'),
   url: requireEnv(process.env.REACT_APP_KEYCLOAK_URL, 'keycloakUrl'),
-} as KeycloakConfig : undefined;
+} as KeycloakConfig : authenticationMode === 'openshift' ? {
```
Comment (ia3andy) [inline packages/launcher-app/src/config.ts]: I guess having a function with `if` to return the `authConfig` would make sense, double ternary operator can be hard to read.

File: packages/launcher-client/src/client/impl/default.launcher.client.ts (hunk 1 above)
Comment (ia3andy) [reply packages/launcher-client/src/client/impl/default.launcher.client.ts]: @gastaldi when are you going to change that?

**Comment by fabric8cd** [comment]
This PR was deployed **with mocked data** in the following URLs for your appreciation:
- Frontend: https://fabric8-launcher-frontend-pr-1231-app.surge.sh
- Welcome App: https://fabric8-launcher-frontend-pr-1231-welcome-app.surge.sh
- Storybook: https://fabric8-launcher-frontend-pr-1231-storybook.surge.sh

**Comment by ia3andy** [comment]
@edewit I spend quite some time making it work, I had a lot of bugs with github. I had to make the github validate token url configurable to be able to test locally. I used `REACT_APP_GITHUB_VALIDATE_URI=https://cors-anywhere.herokuapp.com/https://github.com/login/oauth/access_token` to avoid cors errors locally

**Comment by ia3andy** [comment]
Beside the cors probs I had locally, the other problem was with the fact that once the token is stored, you don't want to call `getGitHubAccessToken` again.

File: packages/launcher-client/src/client/impl/default.launcher.client.ts [hunk 3]
```diff
@@ -165,10 +165,10 @@ export default class DefaultLauncherClient implements LauncherClient {
       headers['X-Git-Provider'] = config.gitProvider;
     }
     if (typeof authorizationToken === 'string') {
-      headers['Authorization'] = `Bearer ${authorizationToken}`;
+      headers['Authorization'] = authorizationToken ? `Bearer ${authorizationToken}` : '';
```
Comment (edewit) [inline packages/launcher-client/src/client/impl/default.launcher.client.ts]: this is bullshit

**Comment by ia3andy** [comment]
got an error "{"error":"invalid_request","error_description":"The request is missing a required parameter, includes an invalid parameter value, includes a parameter more than once, or is otherwise malformed."}"

**Comment by gastaldi** [comment]
@edewit maybe the OpenShift OAuth URL can be returned from the `/api/services/clusters` endpoint? Need to investigate

**Comment by gastaldi** [comment]
@edewit https://github.com/fabric8-launcher/launcher-backend/pull/744

**Comment by ia3andy** [comment]
@gastaldi that may be nice if we make this feature work with multi-cluster. For now it's used for login too (so before loading the cluster list).

**Pull request merged**